vgpu_unlock.patch

```diff
@@ -0,0 +1,152 @@
++ * 2020-12-20 8760c909f54a ("file: Rename __close_fd to close_fd and remove
++ * the files parameter")
+  */
+-#if NV_IS_EXPORT_SYMBOL_PRESENT___close_fd
++#if NV_IS_EXPORT_SYMBOL_PRESENT_close_fd
++    close_fd(fd);
++#elif NV_IS_EXPORT_SYMBOL_PRESENT___close_fd
+     __close_fd(current->files, fd);
+ #else
+     sys_close(fd);
+diff --git kernel/nvidia/nvidia.Kbuild kernel/nvidia/nvidia.Kbuild
+index 8fc929e..193e7b3 100644
+--- kernel/nvidia/nvidia.Kbuild
++++ kernel/nvidia/nvidia.Kbuild
+@@ -173,6 +173,7 @@ NV_CONFTEST_SYMBOL_COMPILE_TESTS += is_export_symbol_present_tegra_dce_client_ip
+ NV_CONFTEST_SYMBOL_COMPILE_TESTS += is_export_symbol_present_dram_clk_to_mc_clk
+ NV_CONFTEST_SYMBOL_COMPILE_TESTS += is_export_symbol_present_get_dram_num_channels
+ NV_CONFTEST_SYMBOL_COMPILE_TESTS += is_export_symbol_present_tegra_dram_types
++NV_CONFTEST_SYMBOL_COMPILE_TESTS += is_export_symbol_present_close_fd
+ 
+ NV_CONFTEST_TYPE_COMPILE_TESTS += acpi_op_remove
+ NV_CONFTEST_TYPE_COMPILE_TESTS += outer_flush_all
+@@ -210,3 +211,6 @@ NV_CONFTEST_GENERIC_COMPILE_TESTS += get_user_pages
+ NV_CONFTEST_GENERIC_COMPILE_TESTS += get_user_pages_remote
+ NV_CONFTEST_GENERIC_COMPILE_TESTS += pm_runtime_available
+ NV_CONFTEST_GENERIC_COMPILE_TESTS += vm_fault_t
++
++# vgpu unlock
++ldflags-y += -T /opt/vgpu_unlock/kern.ld
+diff --git init-scripts/systemd/nvidia-vgpud.service init-scripts/systemd/nvidia-vgpud.service
+index 2da8ff5..7b58195 100644
+--- init-scripts/systemd/nvidia-vgpud.service
++++ init-scripts/systemd/nvidia-vgpud.service
+@@ -13,8 +13,9 @@ Wants=syslog.target
+ 
+ [Service]
+ Type=forking
+-ExecStart=/usr/bin/nvidia-vgpud
++ExecStart=/opt/vgpu_unlock/vgpu_unlock /usr/bin/nvidia-vgpud
+ ExecStopPost=/bin/rm -rf /var/run/nvidia-vgpud
++Environment="__RM_NO_VERSION_CHECK=1"
+ 
+ [Install]
+ WantedBy=multi-user.target
+diff --git init-scripts/systemd/nvidia-vgpu-mgr.service init-scripts/systemd/nvidia-vgpu-mgr.service
+index 24276b3..b50e672 100644
+--- init-scripts/systemd/nvidia-vgpu-mgr.service
++++ init-scripts/systemd/nvidia-vgpu-mgr.service
+@@ -14,8 +14,9 @@ Wants=syslog.target
+ [Service]
+ Type=forking
+ KillMode=process
+-ExecStart=/usr/bin/nvidia-vgpu-mgr
++ExecStart=/opt/vgpu_unlock/vgpu_unlock /usr/bin/nvidia-vgpu-mgr
+ ExecStopPost=/bin/rm -rf /var/run/nvidia-vgpu-mgr
++Environment="__RM_NO_VERSION_CHECK=1"
+ 
+ [Install]
+ WantedBy=multi-user.target
+diff --git kernel/nvidia-vgpu-vfio/nvidia-vgpu-vfio.c kernel/nvidia-vgpu-vfio/nvidia-vgpu-vfio.c
+index a3cc030..22666a2 100644
+--- kernel/nvidia-vgpu-vfio/nvidia-vgpu-vfio.c
++++ kernel/nvidia-vgpu-vfio/nvidia-vgpu-vfio.c
+@@ -2526,14 +2526,18 @@ static irqreturn_t vgpu_msix_handler(int irq, void *arg)
+ 
+     if (pfile && pfile->f_op && pfile->f_op->write)
+     {
+-        old_fs = get_fs();
+-        set_fs(KERNEL_DS);
++        #if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0))
++            old_fs = get_fs();
++            set_fs(KERNEL_DS);
++        #endif
```

Comment:
How does this work? Wouldn't simply removing the set_fs call result in an access violation? Please explain.

Reply:
It has been removed in newer kernel versions (as per [this commit](https://github.com/torvalds/linux/commit/47058bb54b57962b3958a936ddbc59355e4c5504#diff-efc2fe429f2f6e6828b5f8da89c16d371cf1648075dd96f5d85d659610807b1f)), and from what [I've seen](https://github.com/tomaspinho/rtl8821ce/pull/181/files) the patch to deal with that is to simply get rid of those calls.

Reply:
I am so confused.

The commit message for the commit that you linked to is:

> ... this allows the module to compile but MAY NOT work

I feel like the only way that this doesn't cause major issues is if this code never executes, or if the kernel is built without the range checks for the `write` function, which would be a serious security issue.

Reply:
Yeah, I feel like this is a weird workaround, but I have no better idea on how to properly make it work on newer kernel versions.

Reply:
I think the more correct solution is to use `write_iter` instead of `write` when using buffers in kernel space. Ex:
Instead of
```
ret = pfile->f_op->write(pfile, (char *)&val, sizeof(val), &offset);
```
You would want something like
```
struct kvec iov = {
        .iov_base = (void *)&val,
        .iov_len = sizeof(val),
};
struct kiocb kiocb;
struct iov_iter iter;

init_sync_kiocb(&kiocb, pfile);
kiocb.ki_pos = offset;
iov_iter_kvec(&iter, WRITE, &iov, 1, iov.iov_len);
ret = pfile->f_op->write_iter(&kiocb, &iter);
offset = kiocb.ki_pos;
```
Or maybe your solution is correct, anyway I remain confused.

Reply:
```set_fs()``` has been causing security bugs for a long time in the kernel and is now removed starting from kernel 5.10.
To compile any NVIDIA based driver with kernel >= 5.10, the nvidia-vgpu-vfio module needs to patched with IRQ code that removes all the old ```set_fs()``` usage. Patch looks correct. Should also add check for references to ```mm_segment_t```
```
#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0))
   mm_segment_t old_fs
#endif
```

Reply:
I found a plausible explanation for why removing those calls would work in [an LWN.net article](https://lwn.net/Articles/832121/).
> In current kernels, the boundary between kernel and user space is established fairly late in the boot process (but before the init process is started). Before that happens, kernel functions that operate on user-space pointers will happily use kernel-space pointers instead;

So it's possible that at some point these calls were required, but right now the code in question only ever runs early enough that there is no boundary yet. Depending on since when that's the case and what kernels this project realistically supports it's probably even fine to remove them unconditionally.

The article also mentions switching from `write` to `write_iter`, which seems like a more solid fix, if it's worth the effort.